# Лабораторная работа 3 - Нелинейная регрессия
*Широких Артем - АСОиУб-18-1*

## Задание

Для заданных данных провести нелинейную регрессию общего вида для

$f(x) = ax + b/x$

$f_1(x) = ax^2 + bx + c$

$f_3(x) = ae^{-bx} + ab$

In [51]:
class MatrixSize:
    def __init__(self, x: int, y: int):
        self.x = x
        self.y = y

    def __str__(self):
        return '{' + f'x: {self.x}, y: {self.y}' + '}'

class Data:
    def __init__(self, matrix: list):
        self.matrix = matrix
        self.matrix_size = MatrixSize(x=len(self.matrix[0]), y=len(self.matrix))
        self.q = [sum([1 for i in range(self.matrix_size.y) if self.matrix[i][j]]) for j in range(self.matrix_size.x)]


data = Data([
    [0, 87, 83, 77],
    [0, 91, 85, 76],
    [97, 86, 82, 77],
    [92, 88, 84, 79],
    [95, 80, 81, 0],
])

print(data.matrix_size)
print(data.q)

{x: 4, y: 5}
[3, 5, 5, 4]


## Решение

Найдем $N$ общее количество результатов испытаний

In [52]:
n = sum(data.q)
print(n)

17


Определим сумму значений величины $x$ на уровне $A_j$

In [53]:
r = [sum([data.matrix[i][j] for i in range(data.matrix_size.y)]) for j in range(data.matrix_size.x)]
# r[0] = 97
print(r)

[284, 432, 415, 309]


Определим сумму квадратов значений величины $x$ на уровне $A_j$

In [54]:
p = [sum([data.matrix[i][j] ** 2 for i in range(data.matrix_size.y)]) for j in range(data.matrix_size.x)]
# p = [9409, 37380, 34460, 23880]
print(p)

[26898, 37390, 34455, 23875]


Определим факторную и остаточные дисперсии

In [55]:
r_square_divided_q = [(r[j] ** 2) / data.q[j] for j in range(data.matrix_size.x)]

s_fact = (sum(r_square_divided_q) - ((sum(r) ** 2) / n)) / (data.matrix_size.x - 1)
print(s_fact)

s_ost = (sum(p) - sum(r_square_divided_q)) / (data.matrix_size.x - 1)
print(s_ost)

182.9709150326768
30.87222222222287


$S_{fact} \approx 182.971$

$S_{ost} \approx 30.872$

Т.к. $S_{fact} > S_{ost}$, то следует проверить значимость различий между этими дисперсиями.

Вычислим экспериментальное значение критерия.

In [56]:
f_exp = s_fact / s_ost
print(f_exp)

5.926716700716417


$F_{exp} \approx 5.927$

$\alpha = 0.05$

$F_{kr} = qF(\alpha, l, n - 1) = 0.115$

$F_{exp} > F_{kr}$

Т.к. $F_{exp} > F_{kr}$, различие между факторной и остаточной дисперсиями является значимым (при $p=0.05$). В соответствии с методом дисперсионного анализа нулевую гипотезу о равенстве групповых средних следует отвергнуть, т.е. различия между групповыми средними значимы, что соответствует наличию существенного различия между эффективностью воздействия рентгеновского облучения на темп размножения бактерий по задаче.

Вывод: исходя из полученных результатов можно утверждать, что рассматриваемый физический фактор оказыает существенное влияние на размножение бактерий.